src/v/archival/ntp_archiver_service.cc

```diff
@@ -996,14 +1019,18 @@ ss::future<bool> ntp_archiver::maybe_upload_manifest(const char* upload_ctx) {
         co_return false;
     }
 
-    // Do not upload if interval has not elapsed.  The upload loop will call
+    // Do not upload if interval has not elapsed. The upload loop will call
     // us again periodically and we will eventually upload when we pass the
-    // interval.  Make an exception if we are under local storage pressure,
-    // and skip the interval wait in this case.
+    // interval. Make an exception if we are under local storage pressure, or if
+    // a flush operation has completed, and skip the interval wait in these
+    // cases.
+    bool below_manifest_upload_interval
+      = _manifest_upload_interval().has_value()
+        && ss::lowres_clock::now() - _last_manifest_upload_time
+             < _manifest_upload_interval().value();
     if (
-      !local_storage_pressure() && _manifest_upload_interval().has_value()
-      && ss::lowres_clock::now() - _last_manifest_upload_time
-           < _manifest_upload_interval().value()) {
+      (!local_storage_pressure() || !current_flush_complete())
```

Comment:
Given this is a bit tricky to reason about, I'm thinking maybe we should be providing a force option to `maybe_upload_manifest()` and call it around where we upload the topic manifest (along with calling `flush_manifest_clean_offset`)

Reply:
Added `bool force_upload` to `maybe_upload_manifest()` and added a new call where we check `current_flush_complete()` in `upload_until_term_change()`.

I suppose we still want to [check whether the `archival_meta_stm` has been made dirty](https://github.com/redpanda-data/redpanda/blob/d2c41e07ec72e9e87c672b6eac50bd26d4dd9a17/src/v/archival/ntp_archiver_service.cc#L1011) by segment uploads at the beginning of `maybe_upload_manifest()`? Otherwise we could just call `upload_manifest()` directly.